docs/custom-plugins/extended-types.md

```diff
@@ -0,0 +1,74 @@
+# Type extensions in plugins
+
+Redocly CLI in its core has a type tree which defines the structure of the API definition.
+Redocly CLI then uses it to do a type-aware traversal of an OpenAPI definition.
+
+The type tree is built from top level `Types` which can link to child types. For example, here is a [visual reference to the OpenAPI types structure](../../openapi-visual-reference/openapi-node-types.md). You can also check [the code itself](https://github.com/Redocly/redocly-cli/tree/main/packages/core/src/types) for information about specific versions of OpenAPI and other supported document types.
+
+The type tree can be extended by exporting the `typeExtension` function from a custom plugin.
+
+# Type extension example
+
+Consider an OpenAPI that uses an (imaginary) specification extension `x-metadata` which must contain information about the lifecycle state of the API, and which team owns the API. A snippet of the OpenAPI to show the example clearly:
+
+```yaml
+openapi: 3.1.0
+info:
+  title: Food Empire API
+  version: 0.5.1
+  x-metadata:
+    lifecycle: production
+    owner-team: Engineering/Integrations
+```
+
+Define the type extension for the new `x-metadata` section, in this example, we make the `lifecycle` field required.
+
+```js
+const XMetaData = {
+  properties: {
+    lifecycle: { type: 'string'},
+    'owner-team': { type: 'string'},
+  },
+  required: ['lifecycle']
+};
+```
+
+Note the quotes around the `owner-team` key since it contains a hyphen `-`. These fields are strings but you can also define other datastructures if you need to represent something more complex. The [built-in type definitions](https://github.com/Redocly/redocly-cli/tree/main/packages/core/src) are a good place to look for examples.
```

Comment:
```suggestion
Note the quotes around the `owner-team` key since it contains a hyphen `-`. These fields are strings but you can also define other data structures if you need to represent something more complex. The [built-in type definitions](https://github.com/Redocly/redocly-cli/tree/main/packages/core/src) are a good place to look for examples.
```